File: 算法/双指针/num42.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int size = height.size();
        if (size < 3){
            return 0;
        }

        vector<int> left_max(size), right_max(size);

        left_max[0] = height[0];
        for(int i=1; i<size; i++){
            left_max[i] = max(height[i], left_max[i-1]);
        }

        right_max[size-1] = height[size-1];
        for(int i=size-2; i >= 0; i--){
            right_max[i] = max(height[i], right_max[i+1]);
        }

        int ans = 0;
        for(int i=0;i<size;i++){
            ans += min(left_max[i], right_max[i]) - height[i];
        }

        return ans;
    }
};

File: 算法/双指针/num42.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        size = len(height)
        if size < 3:
            return 0

        left = 0
        right = size -1
        left_max = height[left]
        right_max = height[right]
        ans = 0

        while left<right:
            if height[left] < height[right]:
                if height[left] >= left_max:
                    left_max = height[left]
                else:
                    ans += left_max - height[left]
                left += 1
            else:
                if height[right] >= right_max:
                    right_max = height[right]
                else:
                    ans += right_max - height[right]
                right -= 1
        
        return ans